File: Palindrome Linked List/palindrome linked list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        n = self.findLength(head)
        if (n == 1):
            return True
        
        reversePoint = n//2
        if (n%2 == 1):
            reversePoint = (n//2) + 1
        
        head2 = self.findNodeAtPos(head, reversePoint)
        prev = self.findNodeAtPos(head, reversePoint - 1)
        
        self.reverse(head2, prev)
        first = head
        second = prev.next
        
        while (first is not None and second is not None):
            if (first.val != second.val):
                return False
            first = first.next
            second = second.next
        
        return True
    
    def findLength(self, head):
        length = 0
        curr = head
        while (curr is not None):
            length += 1
            curr = curr.next
        return length

    def findNodeAtPos(self, head, pos):
        count = 0
        curr = head
        while (count < pos):
            count += 1
            curr = curr.next
        return curr
    
    def reverse(self, head, prev):
        a = head
        b = head.next
        while (a is not None and b is not None):
            temp = b.next
            b.next = a
            a = b
            b = temp
        head.next = None
        prev.next = a
    

File: Palindrome Linked List/palindrome linked list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        int n = findLength(head);
        if (head.next == null) {
            return true;
        }
        int reversePoint = n/2;
        if (n%2 == 1) {
            reversePoint = (n/2) + 1;
        }
        
        ListNode head2 = findNodeAtPosition(head, reversePoint);
        ListNode prev = findNodeAtPosition(head, reversePoint - 1);
        
        reverse(head2, prev);
        
        ListNode first = head;
        ListNode second = prev.next;
        
        while (first != null && second != null) {
            if (first.val != second.val) {
                return false;
            }
            first = first.next;
            second = second.next;
        }
        
        return true;
    }
    
    
    int findLength(ListNode head) {
        int length = 0;
        ListNode curr = head;
        while (curr != null) {
            length++;
            curr = curr.next;
        }
        return length;
    }
    
    ListNode findNodeAtPosition(ListNode head, int pos) {
        int count = 0;
        ListNode curr = head;
        while (count < pos) {
            count++;
            curr = curr.next;
        }
        return curr;
    }
    
    void reverse(ListNode head, ListNode prev) {
        ListNode a = head;
        ListNode b = head.next;
        
        while (a != null && b != null) {
            ListNode temp = b.next;
            b.next = a;
            a = b;
            b = temp;
        }
        prev.next = a;
        head.next = null;
    }
}